### PR TITLE
Proposed changes to MAM 

### PR DESCRIPTION
**Pull request by annovanvliet**
When evaluating the use of the MAM implementation for usage within our client, I found out there was only support for requesting archived messages to the server. We do use this for requesting archives for Rooms, but for this to work, the MAM requests have to be addressed at the room. therefore I proposed the following changes.

As an extra I included the changes I had to made to be able to compile the code on a windows machine.

### CODE REVIEW

**Comment by Flowdalic** [comment]
Thanks for you contribution, but please *always* create distinct pull request for unrelated things. So please factor the windows thing into a new PR.

**Comment by annovanvliet** [comment]
Removed the windows related PR, and cleaned the empty lines.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/mam/MamManager.java [hunk 1]
```diff
@@ -214,6 +215,33 @@ public MamQueryResult queryArchiveWithEndDate(Date end) throws NoResponseExcepti
...
+     * @param withJid
+     * @param additionalFields
+     * @return the MAM query result
+     * @throws NoResponseException
+     * @throws XMPPErrorException
+     * @throws NotConnectedException
+     * @throws InterruptedException
+     * @throws NotLoggedInException
+     */
+    public MamQueryResult queryArchive(Jid objectJID, String node, Integer max, Date start, Date end, Jid withJid, List<FormField> additionalFields)
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/mam/MamManager.java]: Please change `objectJid` to `mamArchiveAddress`. (BTW: How did you come up with "objectJid"?)

File: smack-experimental/src/main/java/org/jivesoftware/smackx/mam/MamManager.java [hunk 2]
```diff
@@ -289,10 +320,11 @@ private static void addStart(Date start, DataForm dataForm) {
      * @throws InterruptedException
      * @throws NotLoggedInException
      */
-    public MamQueryResult page(DataForm dataForm, RSMSet rsmSet) throws NoResponseException, XMPPErrorException,
+    public MamQueryResult page(Jid objectJID, String node, DataForm dataForm, RSMSet rsmSet) throws NoResponseException, XMPPErrorException,
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/mam/MamManager.java]: Please also keep the `page(DataForm, RSMSet)` method. Smack's 4.2 API is frozen, if this shall go into 4.2 you must not break the API.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/mam/MamManager.java [hunk 3]
```diff
@@ -401,17 +433,21 @@ public MamQueryResult mostRecentPage(Jid chatJid, int max) throws XMPPErrorExcep
     /**
      * Get the form fields supported by the server.
      * 
+     * @param objectJID The object for which archive is requested
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/mam/MamManager.java]: If `objectJID` means what I think it means, then the javadoc makes no sense. It's the address of the MAM archive, isn't it? And also please s/objectJID/mamArchiveAddress/.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/mam/MamManager.java (hunk 1 above)
Comment (annovanvliet) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/mam/MamManager.java]: Changed it to archiveAddress, mam is only implementation detail

About the objectID, Querying messages from an object, but you are right, this not very consistent and not self-explaining

**Comment by annovanvliet** [comment]
Update the API and javadoc

File: smack-experimental/src/main/java/org/jivesoftware/smackx/mam/MamManager.java (hunk 2 above)
Comment (annovanvliet) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/mam/MamManager.java]: kept the original method, also did this for the retrieveFormFields  method

File: smack-experimental/src/main/java/org/jivesoftware/smackx/mam/MamManager.java (hunk 3 above)
Comment (annovanvliet) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/mam/MamManager.java]: tried to explan it better and changed the name

**Comment by Flowdalic** [comment]
Cherry picked as https://github.com/igniterealtime/Smack/commit/c13cddd91a37d1f02573cd5f86572643ffa2a8b0

I made some changes to the API so that it's in sync with the PubSub API, namely one MamManager per archive.

**Pull request closed without merging**